File: blocking/futex/await_blocking_futex.js
//Provides: await_blocking_futex
let await_blocking_futex = 0

//Provides: await_blocking_futex_get
function await_blocking_futex_get() {
  return 0;
}

//Provides: await_blocking_futex_count
//Requires: await_blocking_futex
function await_blocking_futex_count(_t) {
  return await_blocking_futex;
}

//Provides: await_blocking_futex_wait
//Requires: caml_raise_sys_error
//Requires: await_blocking_futex
function await_blocking_futex_wait(_t, count) {
  if (await_blocking_futex === count)
    caml_raise_sys_error("Futex.wait: cannot wait.");
  return await_blocking_futex;
}

//Provides: await_blocking_futex_signal
//Requires: await_blocking_futex
function await_blocking_futex_signal(_t) {
  // Keep the [await_blocking_futex] as a 32-bit integer to ensure that every
  // [await_blocking_futex_signal] call will make it unequal to previous value.
  await_blocking_futex = await_blocking_futex + 1 | 0;
  return 0;
}

File: blocking/futex/await_blocking_futex.ml
(** A handle to a linux futex for waiting on a trigger. *)
type t : immediate

type count : immediate

(** Returns a futex that may be used to wait on a trigger. A new futex may or may not be
    returned by each call to [get]. *)
external get : unit -> t @@ portable = "await_blocking_futex_get"
[@@noalloc]

(** Returns the current count of the futex. *)
external count : t -> count @@ portable = "await_blocking_futex_count"
[@@noalloc]

(** Increments the count of the futex. Any call to [wait] on the same futex will check
    whether the associated trigger has been signaled before suspending the thread. *)
external signal : t -> unit @@ portable = "await_blocking_futex_signal"
[@@noalloc]

(** Wait until the count of the futex changes by suspending the thread until the futex is
    [signal]ed, then return the current count. *)
external wait : t -> count:count -> count @@ portable = "await_blocking_futex_wait"
